native/src/main/java/io/ballerina/stdlib/graphql/runtime/engine/Engine.java

```diff
@@ -161,14 +185,27 @@ public static Object executeQueryResource(Environment environment, BObject conte
         ServiceType serviceType = (ServiceType) TypeUtils.getType(service);
         Type returnType = TypeCreator.createUnionType(PredefinedTypes.TYPE_ANY, PredefinedTypes.TYPE_NULL);
         Object[] arguments = argumentHandler.getArguments();
+
+        Map<String, Object> properties = null;
+        if (ObserveUtils.isObservabilityEnabled()) {
+            if ((getPropertiesToPropagate(environment).get(KEY_INTERCEPTOR) == null ||
+                    !((boolean) getPropertiesToPropagate(environment).get(KEY_INTERCEPTOR))) &&
+                    fieldObject.getBooleanValue(StringUtils.fromString(IS_MAIN_SERVICE))) {
+                properties = new HashMap<>();
+                GraphqlObserverContext observerContext = new GraphqlObserverContext(QUERY_OPERATION);
+                environment.setStrandLocal(KEY_OBSERVER_CONTEXT, observerContext);
+                properties.put(KEY_OBSERVER_CONTEXT, observerContext);
+            } else {
+                properties = getPropertiesToPropagate(environment);
+            }
```

Comment:
This logic seems to be duplicated in multiple places. Shall we move it to a separate method?

Reply:
Sure I will move it to a separate method.